src/acceptor.rs

```diff
@@ -0,0 +1,676 @@
+            }
+            let callback: ReadCallback = try_callback!(callback);
+
+            let mut reader = CallbackReader { callback, userdata };
+
+            let n_read: usize = match acceptor.read_tls(&mut reader) {
+                Ok(n) => n,
+                Err(e) => return rustls_io_result(e.raw_os_error().unwrap_or(EIO)),
+            };
+            unsafe {
+                *out_n = n_read;
+            }
+
+            rustls_io_result(0)
+        }
+    }
+
+    /// Parse all TLS bytes read so far.  If those bytes make up a ClientHello,
+    /// create a rustls_accepted from them.
+    ///
+    /// Parameters:
+    ///
+    /// acceptor: The rustls_acceptor to access.
+    /// out_accepted: An output parameter. The pointed-to pointer will be set
+    ///   to a new rustls_accepted only when the function returns
+    ///   RUSTLS_RESULT_OK. The memory is owned by the caller and must eventually
+    ///   be freed.
+    ///
+    /// Returns:
+    ///
+    /// - RUSTLS_RESULT_OK: a ClientHello has successfully been parsed.
+    ///   A pointer to a newly allocated rustls_accepted has been written to
+    ///   *out_accepted.
+    /// - RUSTLS_RESULT_NOT_READY: a full ClientHello has not yet been read.
+    ///   Read more TLS bytes to continue.
+    /// - Any other rustls_result: the TLS bytes read so far cannot be parsed
+    ///   as a ClientHello, and reading additional bytes won't help.
+    ///
+    /// Memory and lifetimes:
+    ///
+    /// After this method returns RUSTLS_RESULT_OK, `acceptor` is
+    /// still allocated and valid. It needs to be freed regardless of success
+    /// or failure of this function.
+    ///
+    /// Calling `rustls_acceptor_accept()` multiple times on the same
+    /// `rustls_acceptor` is acceptable from a memory perspective but pointless
+    /// from a protocol perspective.
+    #[no_mangle]
+    pub extern "C" fn rustls_acceptor_accept(
+        acceptor: *mut rustls_acceptor,
+        out_accepted: *mut *mut rustls_accepted,
+    ) -> rustls_result {
+        ffi_panic_boundary! {
+            let acceptor: &mut Acceptor = try_mut_from_ptr!(acceptor);
+            if out_accepted.is_null() {
+                return NullParameter
+            }
+            match acceptor.accept() {
+                Ok(None) => rustls_result::NotReady,
+                Err(e) => map_error(e),
+                Ok(Some(accepted)) => {
+                    BoxCastPtr::set_mut_ptr(out_accepted, Some(accepted));
+                    rustls_result::Ok
+                }
+            }
+        }
+    }
+}
+
+impl rustls_accepted {
+    /// Get the server name indication (SNI) from the ClientHello.
+    ///
+    /// Parameters:
+    ///
+    /// accepted: The rustls_accepted to access.
+    ///
+    /// Returns:
+    ///
+    /// A rustls_str containing the SNI field. The returned value is valid
+    /// until the next time a method is called on the `rustls_accepted`.
```

Comment:
I think this is unnecessarily restrictive? Most of the methods on `rustls_accepted` do not mutate, so having pointers into all of the server name, the sig scheme and the cipher suites should be fine.